### ISSUE TITLE
BIG: ability to add services to a running configuration

### ISSUE THREAD

**Issue by rjrodger** [apparatus/fuge#22]
This would be very awesome.

Typically when building a microservice system I leave existing services running and just add new ones for now functionality.

At the moment I have to bring whole fuge system down and restart to do this, which sort of defeats the purpose.

I'm guessing a dynamic reparse and compare against in-memory state should be enough?

**Comment by mcollina** [comment]
So, you want the live-reload also for the fuge config? I think that is doable, but likely it would automate the down/up cycle. Will that be enough?

As a side note, Fuge does not have the advanced automatic analysis that nscale had and that make things way simpler.

**Comment by rjrodger** [comment]
not even automate the up down cycle - I want my existing running services to stay running - just like a live real system :)

**Comment by mcollina** [comment]
I fear that can cause issues and leave everything in a weird situation. Basically you would need to detect any change in any service of the docker-compose.yml.

This is possible, but prone to errors. I think we should split this in two tasks: 1) detect the compose file has changed and reload everything and 2) applying some smartness to the process and only detect and restart what is needed: if the detection thinks it's too complicated (adding a new service is just one thing), it should just restart everything.

BTW, I think adding new services is not the most common tasks, it would occur every 100s of changes you do to a system. However, it is a fairly common problem when bootstrapping a new project.

**Comment by pelger** [comment]
Actually I think this should be fairly straight forward - thanks for flagging will take a look once bandwidth allows

**Comment by AdrianRossouw** [comment]
Just backing up @mcollina here.


i ran into annoying things with nscale before, because it needs to be able to detect changes in environment variables and the like to trigger restarts of already running services.

Otherwise you need to babysit individual ones.

**Comment by AdrianRossouw** [comment]
btw: wouldnt the easiest first cut of this just be to watch the docker-compose.yml and restart the whole system whenever it changes.

it should hit most requirements, work consistently, be simple to implement, and doesn't feel like premature optimisation.

**Comment by rjrodger** [comment]
So I'm going to take a user advocacy perspective here - just because a feature is hard to implement does  not mean it's use case is invalid :)

@mcollina a team building with microservices will be adding services on a weekly basis - greater and greater specializations - yes, this is a little hard-core, but it is the end game. The addition of a single service instance to a live system is the fundamental unit of deployment.

**Comment by mcdonnelldean** [comment]
As a user I would like VERY much to have this :D

**Comment by mcollina** [comment]
@rjrodger yes, I agree. This is why I'm flagging this as a "once or twice a day" operation during normal development. It is annoying though, so we need to address it.

@AdrianRossouw I flagged the stop/start approach as option 1 https://github.com/apparatus/fuge/issues/22#issuecomment-179126579 for a first cut.

**Comment by pelger** [comment]
Stop / start would be a pain for users I think. Auto adding a service though detecting changes in the compose yml may also prove to be confusing. Perhaps the best approach would be to have a compile command in the shell this would recompile the configuration, detect any additional services and bring them in accordingly - thoughts @rjrodger @AdrianRossouw @mcdonnelldean @mcollina

**Comment by mcollina** [comment]
stop / start would probably don't matter, given that a service has a low startup time. We already have change detection for services. So users should be ok in them restarting automatically.

The easiest thing to add is to have a "reload" command that reloads the compose file and restart everything.

Having implemented the change detection system in nscale, I think there are a fairly high number of edge cases that can give people headaches. stop/start is the safest solution, it avoid leaking stuff and so on.

Later, we can handle the case of "adding a new service", which has very little side effect and can be done in place. Changing the config of an already existing service is more tricky. The way I suggest this to be implemented is in steps, adding more "smartness" to a dumb solution (stop/start) as long as we feel comfortable with.

**Comment by mcdonnelldean** [comment]
So here's my issue, it's not a huge one but it's still an issue.

I'd like to be able to leave the working system, as is, running on my machine, right now if I change something within a service that is known it will restart that service, I can also take it off line and spin it up again, all in fuge. This is amazing stuff but it falls down as soon as I need to add new services or infrastructure.

I'm now at the point where I have 4 new microservices to either build or split out from the existing system. Each time I do this exercise I have to take down all my infrastructure and running services. I could of course run my infrastructure in a different terminal but that's a self defeating point.

I love the idea of at least moving to a recompile command as this is pretty much what I am looking for. Having to kill the system because I changed the config seems like bringing a chainsaw to a knife fight in my eyes. 

The nice thing is it sort of self solves the issue of edge cases. The user decides when it is safe to bring things online by triggering the step. If that causes weirdness in their system they can simply stop affected systems and try the compile again.

**Comment by AdrianRossouw** [comment]
I'm speaking from experience with nscale that more complex detection of what to stop/start can become very error prone, incredibly confusing and waste a lot of time when debugging.

It becomes a hunt for 'why didn't this service restart / redeploy', which ends up in a hunt through the fuge/nscale code itself.

Thing is that this logic made sense in the nscale deployment scenario, since these were meant to be live services. Fuge is running on the developer's desktop, and shouldn't really need that kind of complexity.

In an environment where you are probably watching to restart individual services on edits to them, you should be able to restart them as you need. And from what I can see, bringing the whole thing up/down is most likely going to solve this problem for the majority of use cases.

I suspect that implementing @mcollina's first step is going to make the subsequent steps less desirable because the problem will be 99% resolved.

Now, the one edge case here is if you are developing stateful microservices, specifically ones that are only using something like the seneca-mem-store ( @mcdonnelldean ... ). These services are the only ones that can't survive a restart, because they throw away all their data.

The correct answer here would be to stick some kind of data store on them (even if it is a jsonfile store).
I feel that this is a bug in the app, and not a feature requirement for fuge. We shouldn't try to optimize for broken code.

**Comment by mcdonnelldean** [comment]
@AdrianRossouw This doesn't answer my actual need though. To add a new service to the mix I need to stop the whole thing, it's not a big need granted but this does things like stops running calcs in influx. I'm advocating for a recompile as @pelger proposed.

**Comment by AdrianRossouw** [comment]
point taken.

**Comment by rjrodger** [comment]
`reload` === `kill -HUP $PID` just like in the good old days :)
